### DISCUSSION TITLE
Modularity & minimalistic builds - lightweight renderers

### DISCUSSION THREAD

**Discussion by klokan** [maplibre/maplibre#159] [Cross-project feature proposals]
Prepare process for smaller builds of the MapLibre library - defined at compile time or via tree-shaking and dependencies - for removing unused or unwanted functionality - with maximally reduced size and optimized for tight integration with 3rd party libraries. Things like font handling with local platform only,  reduced or even removed available interaction (and rendering cycle), tree shaking optimizations, compile-time disabled complex logic for specific features such as 3d terrain, heatmap, client-side hill shading, etc.

Goal: Web as well as native applications - where size and speed matters a lot - renderering functionality embedded in Leaflet & OpenLayers.

**Ideas / tickets towards this direction should be proposed in different CORE projects of MapLibre.**

Ported from https://roadmap.maplibre.org/c/113-light-renderer-for-leaflet-openlayers

**Comment by fr-an-k** [comment]
It should also be easy to replace parts of the library, i.e. plugins.

It would be great if the rendering system (buckets) would be separated out and made friendly as an API on its own.

Geojson/vector tiles have been innovative but are terrible choices for more powerful/interactive CAD/data visualization, and curves. OffscreenCanvas support could also be more easily added I hope.

Mapbox could never hurt their bottom line like this, but this should be where Maplibre shines, in my opinion, and make the other bounty projects easier.

**Reply by birkskyum** [reply]
@fr-an-k , just out of interest, can you elaborate a bit on what would be better alternatives to Geojson/vector tiles?

**Reply by fr-an-k** [reply]
Geojson is very inefficient for more demanding applications; we have memory issues on older laptops.
GML/WKB/WFS would be good to (incrementally) support because of its use in traditional GIS systems, such as government maps.

Vector tiles can be made non-redundant to save space while still loading incrementally, they don't preserve references within attributes, it's difficult to query all features regardless of being shown at the current zoom level, and they are not suited for quick modifications or working with attributes in more complex ways, because they are not suited for direct modification.

In our case we want 3D pipes & cables because it's much more intuitive than flat lines, but they should still be easy to modify akin to geojson rather than mesh-based files.

What I mean is basically splitting maplibre into standalone packages:
- Basic map functionality (camera and such), but modular, so you can customize input and camera, or project it more easily within another 3D application.
- Bucket types, all standalone so you can easily fork/extend a new geometry/styling type or prune the ones you don't use.
- Source types (for the same reason), so you can use vector tiles or your own type that suits your application. This is already possible to some extend, but at least the bucket system should be documented and easy to use with something other than vector tiles.
- Feature rendering management system that's more flexible than layers, we have ~1000 layers for a project just to handle all the styling and be able to bring the objects of a certain company to the front.

Additionally, it would be nice if you could easily transform the tile vertices (using a shader), for more sophisticated coordinate systems and non-elliptic worlds.
There is no reason to limit maplibre to planets; there are many other use cases for navigating large amounts of visual data, such as microscopic cell biology and star systems.
